src/core/net/nat64_translator.cpp

```diff
@@ -0,0 +1,446 @@
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file includes implementation for the NAT64 translator.
+ *
+ */
+
+#include "nat64_translator.hpp"
+
+#if OPENTHREAD_CONFIG_NAT64_TRANSLATOR_ENABLE
+
+#include "common/code_utils.hpp"
+#include "common/locator_getters.hpp"
+#include "common/log.hpp"
+#include "net/checksum.hpp"
+#include "net/ip4_types.hpp"
+#include "net/ip6.hpp"
+
+namespace ot {
+namespace Nat64 {
+
+RegisterLogModule("Nat64");
+
+Translator::Translator(Instance &aInstance)
+    : InstanceLocator(aInstance)
+{
+    mNat64Prefix.Clear();
+    mIp4Cidr.Clear();
+}
+
+Translator::Result Translator::TranslateFromIp6(Message &aMessage)
+{
+    Result          res = kDrop;
+    Ip6::Header     ip6Header;
+    Ip4::Header     ip4Header;
+    AddressMapping *mapping = nullptr;
+
+    // ParseFrom will do basic checks for the message, including the packet length and IP protocol version.
+    if (ip6Header.ParseFrom(aMessage) != kErrorNone)
+    {
+        LogWarn("outgoing packet is not a valid IPv6 packet, drop");
+        ExitNow(res = kDrop);
+    }
+
+    if (!mNat64Prefix.IsValidNat64() || !ip6Header.GetDestination().MatchesPrefix(mNat64Prefix))
+    {
+        ExitNow(res = kNotTranslated);
+    }
+
+    if (mIp4Cidr.mLength == 0)
+    {
+        // The NAT64 translation is bypassed (will be handled externally)
+        LogDebg("no IPv4 CIDR for NAT64 configured, forward to upper layer");
+        ExitNow(res = kForward);
+    }
+
+    if (ip6Header.GetHopLimit() <= 1)
+    {
+        LogDebg("outgoing packet hop limit reached, drop");
+        ExitNow(res = kDrop);
+    }
+    ip6Header.SetHopLimit(ip6Header.GetHopLimit() - 1);
+
+    mapping = FindOrAllocateMapping(ip6Header.GetSource());
+    if (mapping == nullptr)
+    {
+        LogWarn("failed to get a mapping for %s (mapping pool full?)", ip6Header.GetSource().ToString().AsCString());
+        ExitNow(res = kDrop);
+    }
+
+    aMessage.RemoveHeader(sizeof(Ip6::Header));
+
+    ip4Header.Clear();
+    ip4Header.InitVersionIhl();
+    ip4Header.SetSource(mapping->mIp4);
+    ip4Header.GetDestination().ExtractFromIp6Address(mNat64Prefix.mLength, ip6Header.GetDestination());
+    ip4Header.SetTtl(ip6Header.GetHopLimit());
+    ip4Header.SetIdentification(0);
+
+    switch (ip6Header.GetNextHeader())
+    {
+    case Ip6::kProtoUdp:
+        ip4Header.SetProtocol(Ip4::kProtoUdp);
+        res = kForward;
+        break;
+    case Ip6::kProtoTcp:
+        ip4Header.SetProtocol(Ip4::kProtoTcp);
+        res = kForward;
+        break;
+    case Ip6::kProtoIcmp6:
+        ip4Header.SetProtocol(Ip4::kProtoIcmp);
+        SuccessOrExit(TranslateIcmp6(aMessage));
+        res = kForward;
+        break;
+    default:
+        ExitNow(res = kDrop);
+    }
+
+    {
```

Comment:
Can remove `{` here?

Reply:
Github mismatched the change, this line is removed actually.